File: Sources/ImageHub/Services/OfficeDeploymentTool.swift
import Foundation

/// Gets hold of the Office Deployment Tool's `setup.exe` so nobody has to.
///
/// The alternative was making every operator find the ODT on Microsoft's download
/// centre, run the self-extractor, and remember where they put the result —
/// before their first build could even start.
///
/// It is fetched from Microsoft's own CDN rather than committed to this repo and
/// bundled like wimlib. Partly size, mostly licence: the ODT ships under Microsoft
/// terms that are not ImageHub's to redistribute under, and downloading it from
/// source at build time sidesteps the question entirely. Office itself is never
/// bundled for the same reason, more emphatically — it is a licensed product.
enum OfficeDeploymentTool {
    /// Microsoft's evergreen copy. Versioned download-centre URLs rot every
    /// release; this one is the Click-to-Run bootstrapper Microsoft themselves
    /// point deployment tooling at, and it is the same binary the ODT extracts.
    static let downloadURL = URL(string: "https://officecdn.microsoft.com/pr/wsus/setup.exe")!

    /// Cached beside the rest of ImageHub's support files, so it is fetched once
    /// per Mac rather than once per build.
    static var cachedURL: URL {
        let support = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)
            .first ?? FileManager.default.temporaryDirectory
        return support
            .appendingPathComponent("ImageHub", isDirectory: true)
            .appendingPathComponent("OfficeDeploymentTool", isDirectory: true)
            .appendingPathComponent("setup.exe")
    }

    static var isCached: Bool {
        FileManager.default.fileExists(atPath: cachedURL.path)
    }

    struct ToolError: LocalizedError {
        let message: String
        var errorDescription: String? { message }
    }

    /// Returns a usable `setup.exe`, downloading it if this Mac has not got one.
    ///
    /// `preferred` is the operator's own copy when a template names one; it wins,
    /// because somebody who pinned a version meant it.
    static func resolve(
        preferred: String,
        log: @escaping @Sendable (String) -> Void
    ) async throws -> URL {
        if !preferred.isEmpty {
            let chosen = URL(fileURLWithPath: preferred)
            guard FileManager.default.fileExists(atPath: chosen.path) else {
                throw ToolError(
                    message: "The Office Deployment Tool this template points at is missing: \(preferred)"
                )
            }
            return chosen
        }

        let destination = cachedURL
        if FileManager.default.fileExists(atPath: destination.path) {
            log("Using the cached Office Deployment Tool.")
            return destination
        }

        log("Downloading the Office Deployment Tool from Microsoft…")
        try FileManager.default.createDirectory(
            at: destination.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )

        // Downloaded to a sibling and moved into place, so an interrupted download
        // never leaves a truncated setup.exe looking like a cached one.
        let partial = destination.appendingPathExtension("partial")
        try? FileManager.default.removeItem(at: partial)
        do {
            let (temporary, response) = try await URLSession.shared.download(from: downloadURL)
            if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
                throw ToolError(
                    message: "Microsoft's CDN returned HTTP \(http.statusCode) for the Deployment Tool."
                )
            }
            try FileManager.default.moveItem(at: temporary, to: partial)
        } catch let error as ToolError {
            throw error
        } catch {
            throw ToolError(
                message: """
                    Couldn't download the Office Deployment Tool: \(error.localizedDescription). \
                    Choose a setup.exe by hand on the Apps tab, or check this Mac's connection.
                    """
            )
        }

        // A 7 MB binary that arrives far smaller is a captive-portal login page or
        // an error document, and it would fail much later and less clearly.
        let size = (try? partial.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0
        guard size > 1_000_000 else {
            try? FileManager.default.removeItem(at: partial)
            throw ToolError(
                message: """
                    The Office Deployment Tool downloaded as only \(Int64(size).byteSize), which is \
                    not a real setup.exe — usually a captive portal or a proxy error page. \
                    Check this Mac's connection, or choose a setup.exe by hand on the Apps tab.
                    """
            )
        }

        try? FileManager.default.removeItem(at: destination)
        try FileManager.default.moveItem(at: partial, to: destination)
        log("Office Deployment Tool ready (\(Int64(size).byteSize)).")
        return destination
    }
}

File: Windows/App/Services/OfficeDeploymentTool.cs
using System;
using System.IO;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using ImageHub.Models;
using ImageHub.Support;

namespace ImageHub.Services;

/// <summary>
/// Gets hold of the Office Deployment Tool's setup.exe so nobody has to.
///
/// The alternative was making every operator find the ODT on Microsoft's download
/// centre, run the self-extractor, and remember where they put the result — before
/// their first build could even start.
///
/// It is fetched from Microsoft's own CDN rather than committed to this repo and
/// shipped inside ImageHub. Partly size, mostly licence: the ODT ships under
/// Microsoft terms that are not ImageHub's to redistribute under, and downloading it
/// from source at build time sidesteps the question entirely. Office itself is never
/// bundled for the same reason, more emphatically — it is a licensed product.
/// </summary>
public static class OfficeDeploymentTool
{
    /// <summary>
    /// Microsoft's evergreen copy. Versioned download-centre URLs rot every release;
    /// this one is the Click-to-Run bootstrapper Microsoft themselves point deployment
    /// tooling at, and it is the same binary the ODT extracts.
    /// </summary>
    public const string DownloadUrl = "https://officecdn.microsoft.com/pr/wsus/setup.exe";

    /// <summary>Cached under ImageHub's own folder, so it is fetched once per PC rather than per build.</summary>
    public static string CachedPath => AppPaths.OfficeDeploymentTool;

    public static bool IsCached => File.Exists(CachedPath);

    /// <summary>
    /// Returns a usable setup.exe, downloading it if this PC has not got one.
    /// <paramref name="preferred"/> is the operator's own copy when a template names
    /// one; it wins, because somebody who pinned a version meant it.
    /// </summary>
    public static async Task<string> ResolveAsync(
        string preferred,
        Action<string> log,
        CancellationToken cancellation = default)
    {
        if (preferred.Length > 0)
        {
            if (!File.Exists(preferred))
            {
                throw new BuildException(
                    "The Office Deployment Tool this template points at is missing: " + preferred);
            }
            return preferred;
        }

        string destination = CachedPath;
        if (File.Exists(destination))
        {
            log("Using the cached Office Deployment Tool.");
            return destination;
        }

        log("Downloading the Office Deployment Tool from Microsoft…");
        Directory.CreateDirectory(Path.GetDirectoryName(destination)!);

        string partial = destination + ".partial";
        try { if (File.Exists(partial)) { File.Delete(partial); } } catch (IOException) { }

        try
        {
            using HttpResponseMessage response = await Downloader.Shared
                .GetAsync(DownloadUrl, HttpCompletionOption.ResponseHeadersRead, cancellation)
                .ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
            {
                throw new BuildException(
                    $"Microsoft's CDN returned HTTP {(int)response.StatusCode} for the Deployment Tool.");
            }
            await using (Stream source = await response.Content.ReadAsStreamAsync(cancellation)
                             .ConfigureAwait(false))
            await using (FileStream target = File.Create(partial))
            {
                await source.CopyToAsync(target, cancellation).ConfigureAwait(false);
            }
        }
        catch (BuildException)
        {
            throw;
        }
        catch (OperationCanceledException)
        {
            throw;
        }
        catch (Exception error)
        {
            throw new BuildException(
                $"Couldn't download the Office Deployment Tool: {error.Message}. Choose a setup.exe "
                + "by hand on the Apps tab, or check this PC's connection.");
        }

        // A 7 MB binary that arrives far smaller is a captive-portal login page or an
        // error document, and it would fail much later and less clearly.
        long size = 0;
        try { size = new FileInfo(partial).Length; } catch (IOException) { }
        if (size <= 1_000_000)
        {
            try { File.Delete(partial); } catch (IOException) { }
            throw new BuildException(
                $"The Office Deployment Tool downloaded as only {Formatting.ByteSize(size)}, which is "
                + "not a real setup.exe — usually a captive portal or a proxy error page. Check this "
                + "PC's connection, or choose a setup.exe by hand on the Apps tab.");
        }

        try { if (File.Exists(destination)) { File.Delete(destination); } } catch (IOException) { }
        File.Move(partial, destination);
        log($"Office Deployment Tool ready ({Formatting.ByteSize(size)}).");
        return destination;
    }
}
